packages/table/src/Row.tsx

```diff
@@ -0,0 +1,323 @@
+        }
+
+        if (
+          isComponentType(child, 'Cell') &&
+          child.props.rowSpan > 1 &&
+          !hasRowSpan &&
+          !shouldDispatchHasRowSpan
+        ) {
+          shouldDispatchHasRowSpan = true;
+        }
+      });
+
+      if (shouldDispatchHasNestedRows) {
+        tableDispatch({
+          type: TableTypes.SetHasNestedRows,
+          payload: true,
+        });
+      }
+
+      if (shouldDispatchHasRowSpan) {
+        tableDispatch({
+          type: TableTypes.SetHasRowSpan,
+          payload: true,
+        });
+      }
+    }, [children, hasNestedRows, hasRowSpan, tableDispatch]);
+
+    // Iterating over children twice because generated memoized values have different dependants
+    const renderedChildren = React.useMemo(() => {
+      const Icon = isExpanded ? ChevronDownIcon : ChevronRightIcon;
+      const chevronButton = (
+        <IconButton
+          onClick={() => setIsExpanded(curr => !curr)}
+          aria-label="chevron"
+          className={iconButtonMargin}
+        >
+          <Icon aria-label="chevron" color={uiColors.gray.dark2} />
+        </IconButton>
+      );
+      const renderedChildren: Array<React.ReactElement> = [];
+      let hasSeenRow = false;
+
+      React.Children.forEach(children, (child, index) => {
+        if (isComponentType(child, 'Cell')) {
+          if (!child.props.children) {
+            return null;
+          }
+
+          renderedChildren.push(
+            React.cloneElement(child, {
+              children: (
+                <span className={truncation}>{child.props.children}</span>
+              ),
+              key: `${indexRef.current}-${index}`,
+              disabled: child.props.disabled || disabled,
+            }),
+          );
+        } else {
+          if (isComponentType(child, 'Row')) {
+            hasSeenRow = true;
+          }
+        }
+      });
+
+      if (hasSeenRow) {
+        renderedChildren[0] = React.cloneElement(renderedChildren[0], {
+          children: (
+            <>
+              {chevronButton}
+              <span className={truncation}>
+                {renderedChildren[0].props.children}
+              </span>
+            </>
+          ),
+          className: cx(displayFlex, className),
+          key: `${indexRef.current}-${renderedChildren[0].props.children}`,
+        });
+      }
+
+      return renderedChildren;
+    }, [children, disabled, className, isExpanded, setIsExpanded]);
+
+    // Iterating over children twice because generated memoized values have different dependants
+    const nestedRows = React.useMemo(() => {
+      let hasSeenFirstCell = false;
+      const nestedRows: Array<React.ReactElement> = [];
+
+      React.Children.forEach(children, (child, index) => {
+        if (isComponentType(child, 'Cell') && !hasSeenFirstCell) {
+          hasSeenFirstCell = true;
+        }
+
+        if (child != null && isComponentType(child, 'Row')) {
+          nestedRows.push(
```

Comment:
We're rendering nestedRows optimistically. This means that we render them even if they're not appearing on the screen. This is a big performance hit, as most commonly people don't have all rows expanded.

Instead we need to change this to being lazy. Only render and compute nested rows when they're expanded and visible in the page.

This is absolutely a big area of performance cost, and switching to lazy rendering will be a very short way of fixing that.

Reply:
I think originally we were handling this lazily, but we were unable to make the transition work as we wanted. 

The original idea (which did not work with the current transitions) was to not mount the nested rows when `isExpanded` is false. We can take this a step further by not even calculating the nested rows when `isExpanded` is false, just have to decide if it's worth sacrificing the transition. cc: @DesignerDave

Reply:
You'll need to remind me why the transition was dependent on this.

Reply:
There's nothing for the Transition component to animate if the nested rows haven't yet been calculated

Reply:
That's what it appears at least

Reply:
If we need to prioritize: animation or performance, how do you both feel?